### PR TITLE
[WIP] Add data simulation capabilities leveraging sklearn

### PR DESCRIPTION
**Pull request by gpfreitas**
The idea is to wrap some `sklearn.datasets` functions to add extra functionality.

The reasons to wrap the sklearn routines are:

- We want new types easily (xarray.Dataset, etc.)
- We want multidimensional features.
- We want to generate sample weights together with synthetic data.

Motivating usecases:

- Generate data to test models that predict temperature or humidity at
  3d-spatial coordinates and time based on earlier measurements at nearby/same
  coordinates. (Is that right, @PeterDSteinberg?)

### CODE REVIEW

**Comment by PeterDSteinberg** [comment]
Looks good @gpfreitas .  Here's one idea for passing in the extra arguments we may want in order to customize `astype` (which currently doesn't accept `xarray.Dataset`):

```

def synthetic_coords(shp, layers):
    space = np.arange(shp[0])
    coords = dict(layers=layers, space=space)
    dims = ('space', 'layers')
    return coords, dims


def as_xarray_dataset(xy, is_x=True, **kw):
    layers = kw.get('layers', None)
    if is_x:
        cols = xy.shape[1]
        layers = kw.get('layers', None)
        if layers is None:
            layers = ['label_{}'.format(c) for c in range(cols)]
    else:
        if xy.ndim == 1:
            cols = 1
            xy = xy[:, np.newaxis]
        elif xy.ndim == 2:
            cols = xy.shape[1]
        else:
            raise ValueError('Expected 1D or 2D arg xy')
        if layers is None:
            layers = ['true_y']
    coords, dims = synthetic_coords(xy.shape, layers)
    arrs = {'flat': xr.DataArray(xy, coords=coords, dims=dims)}
    return xr.Dataset(arrs)


def flexible_out(skl_make_data_func):
    """Add keyword arguments to skl_make_data_func.

    Note: a decorator doesn't solve this. It could add the functionality, but
    it would not change the docstring or the signature of the function, as it
    would keep the original one from the wrapped function from sklearn. So we
    have to do the process manually.
    """
    def wrapper(*args, astype=None, type_kw=None, **kwargs):
        type_kw = type_kw or {}
        X, y = skl_make_data_func(*args, **kwargs)
        if astype:
            if astype == xr.Dataset:
                astype = as_xarray_dataset
            Xcast = astype(X, **type_kw)
            ycast = astype(y, is_x=False, **type_kw)
            return Xcast, ycast
        else:
            return X, y
```
The above snippet replaces the top of `flexible_out` function and adds two util funcs for `xarray.Dataset`.  Note that the `sklearn` transformers used by `elm`/`earthio` expect a `X` `Dataset` with a `DataArray` called `flat` and coords that include the keys `space` and `band` (Note `band` is becoming `layer` [See also #173](https://github.com/ContinuumIO/elm/issues/173))

Example usage:

```
In [1]: from xarray_filters.datasets import *

In [2]: X, y = make_blobs(astype=xr.Dataset)

In [3]: X
Out[3]:
<xarray.Dataset>
Dimensions:  (layers: 2, space: 100)
Coordinates:
  * layers   (layers) <U7 'label_0' 'label_1'
  * space    (space) int64 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 ...
Data variables:
    flat     (space, layers) float64 3.69 9.34 -2.145 5.467 -3.389 6.041 ...

In [4]: y
Out[4]:
<xarray.Dataset>
Dimensions:  (layers: 1, space: 100)
Coordinates:
  * layers   (layers) <U6 'true_y'
  * space    (space) int64 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 ...
Data variables:
    flat     (space, layers) int64 1 2 2 2 0 0 0 2 1 1 1 2 0 0 1 1 0 0 2 0 1 ...

```

**Comment by gpfreitas** [comment]
@PeterDSteinberg 

I think I got the final design for this. Take a look at the commit message above (commit 73acb65) for an explanation. The module docstring contains more info.

I added your `to_xarray_....` method. But only in the case where "is_X=True", and assuming X is always 2d. Let's discuss more so that I can understand the various use cases (what do you have in mind for `is_x=False`, for example), write some simple tests, and quickly wrap this up.

Until we meet, I'll move on to other issues.

**Comment by gpfreitas** [comment]
Hmm, I just read again your comments about the specific elm/earthio needs. Let's talk about it.

**Comment by PeterDSteinberg** [comment]
Hi @gpfreitas I like your approach here and how you have wrapped things while maintaining the docstrings of the underlying `sklearn` functions.  I'll take over this PR now and you can focus on the notebook example of [lidar / DEM data processing we discussed - elm issue #186 ](https://github.com/ContinuumIO/elm/issues/186).  I'll change the assignee on this issue to myself, then ping @gbrener to review the PR sometime next week.

**Comment by gpfreitas** [comment]
Cool.

One last remark: the code right now supports data transformations from simulated (X, y). If you decide to accept other inputs (just the feature matrix X, or something else), then I'd encourage changing the class name to something more generic, like `DataConverter`, and create new initializers, like `from_xy`, `from_x`, `from_somethingelse`, and maybe set the default initializer `__init__` to be `from_xy`.

**Comment by PeterDSteinberg** [comment]
Hi @gpfreitas - In the recent commit
 * I did some clean up changes and renaming 
 * added a wrapper around `fetch_lfw_people`, a function from `sklearn.datasets` that returns a 1 to ~45 grayscale images for some well known people, to return them as `DataArray`s
* added TODO notes in code you can address next week

**Comment by gpfreitas** [comment]
@PeterDSteinberg I need some feedback to make this better. The API should be right (or really close to it). You can just call, say, `make_blobs(astype='array')` to get the original sklearn behavior, or `astype=datasets` (which fails without additional arguments).

I also added notes near your TODOs. I'd like to move them out of there to notes in PR, or issues, etc.

If you can give me a list of test cases that should succeed (I can try to infer what the output should be, but just knowing it would not fail would be a good first step), then I can start writing tests and fixing any failures.

**Comment by gpfreitas** [comment]
@PeterDSteinberg 

In particular, any explanation and concrete examples in the docstring of `XyTransformer.to_dataset` would be really great.

File: xarray_filters/datasets.py [hunk 1]
```diff
@@ -1,117 +1,415 @@
...
+        # Step 1: process positional and keyword arguments
+        skl_kwds = skl_argspec.args + skl_argspec.kwonlyargs
+        # splitting arguments into disjoint sets; astype is a special argument
+        skl_kwargs = {k: val for (k, val) in kwargs.items() if k in skl_kwds}
+        astype = kwargs.get('astype', default_astype)
+        type_kwargs = {k: val for (k, val) in kwargs.items() if (k not in
+            skl_kwds and k != 'astype')} 
+
+        # Step 2: obtain the XyTransformer object
+        try:
```
Comment (gpfreitas) [inline xarray_filters/datasets.py]: @PeterDSteinberg 

Now I'm thinking a better option is to just read in the output of the sklearn make_* function and explicitly check if the lenght of the output is 2, then raise an error if that is the case. Yea/Nay?

File: xarray_filters/datasets.py [hunk 2]
```diff
@@ -1,117 +1,415 @@
...
+            print('kwargs', skl_kwargs)
+            X, y = skl_sampler_func(*args, **skl_kwargs)
+            # some skl_sampler_funcs do not return X, y, leading to a ValueError
+        except ValueError as e:
+            if 'values to unpack' in str(e):
+                error_msg = 'Function {} does not return a tuple of 2 elements'
+                raise ValueError(error_msg.format(skl_sampler_func.__name__))
+            else:
+                raise
+        # Assert X is a feature matrix and y has the corresponding labels
```
Comment (gpfreitas) [inline xarray_filters/datasets.py]: Actually maybe just factor out the whole "output shape checking" for the sklearn make_* function into an inner func here.

File: xarray_filters/datasets.py [hunk 3]
```diff
@@ -1,117 +1,415 @@
...
+        dset[clean_name] = data_arr
+    attrs = dict(metadata=[args, kw])
+    # TODO - PR 3's MLDataset instead of Dataset
+    dset = xr.Dataset(dset, attrs=attrs)
+    # TODO - allow converting to dataframe, numpy array?
+    return dset
+
+
+# Convert all sklearn functions that admit conversion
+converted_make_funcs = dict()  # holds converted sklearn funcs
```
Comment (gpfreitas) [inline xarray_filters/datasets.py]: `converted_make_funcs` and `sklearn_make_funcs` should probably be "hidden" with leading underscores.

**Comment by gpfreitas** [comment]
@PeterDSteinberg 

Another noteworthy point: as it is, we can convert most, but not all the of the `sklearn.datasets.make_*` functions. See the `xarray/datasets.py` module docstring to see what we can handle and what we cannot (essentially we assume the `make_*` function returns X, y and that it can be called with default values alone, like `make_blobs()`).

Allowing for `make_*` functions that require arguments but still return `X, y` is an easy fix. Allowing for `make_*` functions that return something that is not `X, y` (like various `make_*_matrix` functions) may be more of a case-by-case fix (not that many cases though, around 5).

**Comment by gpfreitas** [comment]
@PeterDSteinberg , @gbrener (ref. commit 22fc5ab)

With the latest commits, we have a working `XyTransformer.to_dataset` to convert data to `xarray.Dataset` objects.

I also added a `utils.py` file with some utilities lifted from xarray (they were not part of the public API, so I chose to not just their imported versions) and a notebook with example usage.

There are a few limitations in the current code:

- We convert functions from `sklearn.datasets.make_*` _only if_
  - they work without any user-passed arguments (all defaults), _and_;
  - they return a tuple `(X, y)` where `X` is a 2d array of features and `y` is a 1d array of labels.
- `fetch_lfw_people` is not implemented
- `XyTransformer.to_ml_features` is not implemented.
- MLDataset is not incorporated yet.

Some of the current limitations can be addressed as issues later, so I will refrain from doing more work on this until someone reviews what has been done so far.

If you want an overview of the functionality, look in the notebook [`examples/20170905_Nasa_meeting_demo_xarray_filters.ipynb`](https://github.com/ContinuumIO/xarray_filters/blob/22fc5abafc4c13011faa700989c1b9c177ac78b0/xarray_filters/examples/20170905_Nasa_meeting_demo_xarray_filters.ipynb).

I guess our goal is to get this merged as soon as possible, together with PR #3. With that in mind:

- Should we merge this before PR #3? The code here is generic enough that it could be merged into master with little modifications (including adding some tests), and right now it seems orthogonal to what's been done in PR #3.
  - If the answer is "yes", what do we need to merge? I'm guessing add tests, clean up docs, and make sure tests pass in CI (I may need help here).
  - If the answer is "no", then how do we proceed? Do we merge this into PR #3?

**Comment by gpfreitas** [comment]
@gbrener and I decided that the best way forward is to merge this work into PR #3.

**Comment by gpfreitas** [comment]
Merged into PR #3. Remaining work to be continued in PR #3 is to write tests and cleanup docs.
